### PR TITLE
refactor(catalog): replaces opm base image for catalog images with th…

### LINKED ISSUE #274: [RFE] Replace OPM base image image for catalog building with the requested catalog image
To ensure compatibility with the OpenShift cluster version, `oc-mirror` needs to replace the pinned `opm` image that populates the base image for the rebuilt catalog images with the catalog image that is being requested. This will ensure that the version of `opm` contained in the image is compatible with the cluster version.

Refactor Proposal:

1. Store the catalog image in an OCI layout alongside the generated index.json
2. During the publish phase read the OCI Layout into a `v1.Image` (go-containerregsitry) and use it as the `opm` image was used.

### PR DESCRIPTION
**Pull request by jpower432**
…e originall requested catalog image

Closes #274

Signed-off-by: Jennifer Power <barnabei.jennifer@gmail.com>

# Description

This PR refactors the way the catalogs are currently being built by replacing the `opm` base image with the mirrored catalog image being requested to ensure the `opm` version for the FBC is supported. It also adds the heads-only vs package filtering workflow (breaking change). If `headsOnly` is false the catalog will be rebuilt without include the previously created FBC.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by jpower432** [comment]
Converting back to draft while I change to the API version

**Comment by afflom** [comment]
The stated behavior is what we decided. Aside from that, I have to defer to @deejross for the code review.

**Comment by jpower432** [comment]
@jharmison-redhat Can you run an integration test on this branch before I cancel the hold, please?

**Comment by jharmison-redhat** [comment]
Will get to this once I finish [this](https://github.com/jharmison-redhat/oc-mirror-e2e/issues/10) up. Should be tomorrow. Apologies for delays while I was out :)

**Comment by jharmison-redhat** [comment]
Late test notes!

Tested with [integration environment v0.4.11](https://github.com/jharmison-redhat/oc-mirror-e2e).

scenario.yml:
```yaml
---
cluster_name: disco
openshift_version: "4.9"

scenario_stubs:
  metadata_backend: registry
  mirror_method: to_registry
  registry_type: docker_registry

oc_mirror_source:
  repo: https://github.com/openshift/oc-mirror.git
  version: main

imageset_config_override: |
  apiVersion: mirror.openshift.io/v1alpha2
  kind: ImageSetConfiguration
  storageConfig:
    local:
      path: metadata
  mirror:
      ocp:
        channels:
          - name: stable-4.9
            minVersion: '4.9.22'
            maxVersion: '4.9.22'
      operators:
      - catalog: registry.internal.disco.redhat4govaws.io/catalog/redhat-operator-index:v4.9
        headsOnly: false # References latest version of each operator in catalog (true is the default value and can be omitted)
        packages:
          - name: sriov-network-operator
            channels:
              - name: '4.9'
          - name: cluster-logging
            channels:
              - name: 'stable'
```

1. `make run ANSIBLE_PLAYBOOKS=create`
2. `skopeo login registry.disco.redhat4govaws.io --tls-verify=false -u admin -p $(cat example/output/registry_admin_password)`
3. `skopeo copy --remove-signatures --dest-tls-verify=false docker://registry.redhat.io/redhat/redhat-operator-index:v4.9-1646041330 docker://registry.disco.redhat4govaws.io/catalog/redhat-operator-index:v4.9`
4. `make run ANSIBLE_PLAYBOOKS=test`
5. ssh to bastion
6.
  ```console
  $ export KUBECONFIG=~/install/auth/kubeconfig
  $ oc get packagemanifest cluster-logging -ojsonpath='{.status.channels[0].currentCSV}'; echo
  cluster-logging.5.2.7-18
  $ exit
  ```
7. `skopeo copy --remove-signatures --dest-tls-verify=false docker://registry.redhat.io/redhat/redhat-operator-index:v4.9 docker://registry.disco.redhat4govaws.io/catalog/redhat-operator-index:v4.9`
8. ssh to registry
9.
  ```console
  $ cd mirror-data
  $ ~/oc-mirror/bin/oc-mirror --config ~/imageset-config.yml docker://registry.internal.disco.redhat4govaws.io/oc-mirror
  <-- much mirroring output -->
  INFO Wrote CatalogSource manifests to oc-mirror-workspace/results-1647638730
  INFO Wrote ICSP manifests to oc-mirror-workspace/results-1647638730
  $ exit
  ```
10. ssh back to bastion
11.
  ```console
  $ export KUBECONFIG=~/install/auth/kubeconfig
  $ oc delete pod -l olm.catalogSource=redhat-operator-index -n openshift-marketplace
  pod "redhat-operator-index-75lpm" deleted
  $ oc get pod -n openshift-marketplace
  NAME                                   READY   STATUS    RESTARTS      AGE
  marketplace-operator-fbd7c964f-xftg5   1/1     Running   4 (30m ago)   39m
  redhat-operator-index-wj9hn            1/1     Running   0             47s
  $ oc get packagemanifest cluster-logging -ojsonpath='{.status.channels[0].currentCSV}'; echo
  cluster-logging.5.3.5-20
  ```

tl;dr once again, lgtm
